### PR TITLE
feat: add physical devices support

### LINKED ISSUE #121: New Feature: Display connected devices to allow copying name, udid, or pasting to Android device
Hey,

Would you be interested in also listing connected iOS, Android devices?

A user can then easily know if a device is successfully connected, copy name/udid, and also paste into the android device in case you have text on your computer and you want to paste it into the device

### PR DESCRIPTION
**Pull request by itsramiel**
### Display physical iOS and Android devices

closes #121 
The idea here to be able to view connected devices.

For iOS, devices are visible even if not connected via cable. The system also marks the device as offline if the device is not connected to the internet and not connected by cable, otherwise it is marked as online. The booted property for iOS physical devices is determined according whether the device is online or not.

For android, connected devices are figured out using `adb devices -l`

Applescript commands are adjusted to account for the changes.
[Link to raycast extension pr](https://github.com/raycast/extensions/pull/14229)

### Motivation
For iOS, this allows you to copy your uuid which can be useful at times or you create a custom command.
For android, you can paste to clipboard(main reason I worked on this pr), launch logcat, and also custom commands.

#### Other changes
I hide sections that do not include any devices which wasnt the case previously. Assuming you are ios dev only that work only with simulators, you would see 3 other empty sections, which I personally think is not nice but that is just my opinion

## Changelog:

feat: display physical ios and android devices
feat: hide sections that include no devices

## Test Plan:

Android

https://github.com/user-attachments/assets/15236eaa-7cd8-4e48-90f2-094f4cc8a9bb

iOS(didnt record video cause would like to keep the uuid private 🤓):
image

### CODE REVIEW

File: MiniSim/Service/DeviceService.swift [hunk 1]
```diff
@@ -263,14 +307,33 @@ extension DeviceService {
...
 
+  static func getIOSPhysicalDevices() throws -> [Device] {
+    let devicesOutput = try shellOut(
+      to: ProcessPaths.xcrun.rawValue,
+      arguments: ["xctrace", "list", "devices"]
+    )
+    let splitted = devicesOutput.components(separatedBy: "\n")
+
+    let currentDeviceUuid = try shellOut(
+      to: "system_profiler SPHardwareDataType | awk '/UUID/ { print $3; }'"
```
Comment (itsramiel) [inline MiniSim/Service/DeviceService.swift:323]: remove dependency on awk

**Comment by itsramiel** [comment]
@okwasniewski I believe it is ready

**Comment by okwasniewski** [comment]
Hey @itsramiel, 

Thanks for this PR! 

I'm going to review and test it over the weekend 🙌

File: MiniSim/MenuItems/DeviceListSection.swift [hunk 2]
```diff
@@ -8,15 +8,21 @@
...
+    case androidPhysical
+    case androidVirtual
 
     var title: String {
         switch self {
-        case .iOS:
+        case .iOSPhysical:
+            return NSLocalizedString("iOS Devices", comment: "")
+        case .iOSVirtual:
             return NSLocalizedString("iOS Simulator", comment: "")
```
Comment (okwasniewski) [inline MiniSim/MenuItems/DeviceListSection.swift:21]: Let's make every menu section in the plural form "iOS Simulators", "Android Simulators"

File: MiniSim/MenuItems/SubMenuItem.swift [hunk 3]
```diff
@@ -145,26 +145,48 @@ enum SubMenuItems {
...
         Separator(),
 
-        ColdBoot(),
-        NoAudio(),
         ToggleA11y(),
         Paste(),
-        DeleteEmulator(),
         LaunchLogCat()
-    ]
+      ]
```
Comment (okwasniewski) [inline MiniSim/MenuItems/SubMenuItem.swift:174]: Some unnecessary diff

File: MiniSim/Service/DeviceService.swift [hunk 4]
```diff
@@ -215,7 +216,50 @@ class DeviceService: DeviceServiceProtocol {
...
 extension DeviceService {
-  private static func parseIOSDevices(result: [String]) -> [Device] {
+  private static func parseIOSPhysicalDevices(result: [String]) -> [Device] {
+    var devices = [Device]()
+
+    let deviceNameIdx = 1
+    let osVersionIdx = 2
+    let identifierIdx = 3
+    var isDevice = false
+    var isOnline = true
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:226]: Should we default to the device being online?

File: MiniSim/Service/DeviceService.swift [hunk 5]
```diff
@@ -215,7 +216,50 @@ class DeviceService: DeviceServiceProtocol {
...
+      if let match = line.match("== ([\\s\\w]+) ==").first {
+        isDevice = match[1].contains("Devices")
+        isOnline = !match[1].contains("Offline")
+        continue
+      }
+
+      guard isDevice else { continue }
+
+      // The version is set as optional which only is the case for the device running
+      // the application, for exmaple your macbook, maybe we can make it not optional
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:237]:
```suggestion
      // the application, for example your MacBook, maybe we can make it not optional
```

File: MiniSim/Service/DeviceService.swift [hunk 6]
```diff
@@ -389,7 +472,40 @@ extension DeviceService {
...
+    let devices = try getAndroidPhysicalDevices()
+    return emulators + devices
+  }
+
+  static func getAndroidPhysicalDevices() throws -> [Device] {
+      let adbPath = try ADB.getAdbPath()
+      let output = try shellOut(to: adbPath, arguments: ["devices", "-l"])
+      var splitted = output.components(separatedBy: "\n")
+      splitted.removeFirst() // removes 'List of devices attached'
+      let filtered = splitted.filter { !$0.hasPrefix("emulator-") }
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:485]: Maybe we can filter this out instead of always assuming this is at the first place?

**Review by okwasniewski — COMMENTED** [review]
Overall looks good! 

I did a first code pass, going to test it more over the weekend

File: MiniSim/MenuItems/DeviceListSection.swift (hunk 2 above)
Comment (itsramiel) [reply MiniSim/MenuItems/DeviceListSection.swift:21]: Sounds like a good idea. What about also changing `Android Simulator` to `Android Emulators`? You also use `emulator` in raycast

File: MiniSim/MenuItems/SubMenuItem.swift (hunk 3 above)
Comment (itsramiel) [reply MiniSim/MenuItems/SubMenuItem.swift:174]: Is it? I think the indentation changes because of the switch scope or?

File: MiniSim/Service/DeviceService.swift (hunk 4 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:226]: I personally think it doesnt matter if I set it to true or false. I would rather not set a value but since `isOnline` is set only in the `if let`:
https://github.com/okwasniewski/MiniSim/blob/f60b1bd7991919147ac8ad9102575b01734f5844/MiniSim/Service/DeviceService.swift#L228-L232

then swift will complain that it may not be set and I would not want to make it optional, thats why I add an initial value

The reason I think it doesnt matter is because the value will be set again according to:
https://github.com/okwasniewski/MiniSim/blob/f60b1bd7991919147ac8ad9102575b01734f5844/MiniSim/Service/DeviceService.swift#L230

But I am open to another approach

File: MiniSim/Service/DeviceService.swift (hunk 5 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift]: Done

File: MiniSim/Service/DeviceService.swift (hunk 6 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:485]:
```suggestion
      splitted.removeFirst() // removes 'List of devices attached'
      let filtered = splitted.filter { !$0.contains("emulator") }
```
Like this? Yeah thats fine for me too

File: MiniSim/Service/DeviceService.swift [hunk 7]
```diff
@@ -41,6 +41,7 @@ class DeviceService: DeviceServiceProtocol {
   private enum ProcessPaths: String {
     case xcrun = "/usr/bin/xcrun"
     case xcodeSelect = "/usr/bin/xcode-select"
+    case systemProfiler = "/usr/sbin/system_profiler"
```
[the remark this answers is not in the document]
Comment (okwasniewski) [reply MiniSim/Service/DeviceService.swift:44]: I just checked and it looks like system_profiler can be executed without sudo if you don't expect full output. If the base output (without sudo) is enough for our use case then let's leave it like this. 

Btw what do you think about creating a small abstraction on top of the systemProfiler utility? 

Maybe a small separate class that encapsulates the logic? 

Like `SystemProfiler.getCurrentUUID()`

**Comment by okwasniewski** [comment]
I've compiled the app and for some reason I can see my macbook there: 

Here is the output of `xctrace list devices`

```
❯ xctrace list devices
== Devices ==
MacBook Pro (Oskar) (123)
iPhone (Oskar) (17.6.1) (123)

== Devices Offline ==
Apple Vision Pro (2.0) (123)
```

Screenshot 2024-08-31 at 14 48 15

It would be nice to support the naming convention of `iPhone (Name)` as this is the default naming in my country and probably others too.

**Comment by okwasniewski** [comment]
Also I found this: `xcrun devicectl list devices` which shows the output for listing only devices. Maybe faster (as it doesn't list simulators)

**Comment by okwasniewski** [comment]
Looks like Android Devices are working smoothly!

Screenshot 2024-08-31 at 16 38 38

File: MiniSim/MenuItems/DeviceListSection.swift (hunk 2 above)
Comment (itsramiel) [reply MiniSim/MenuItems/DeviceListSection.swift:21]: Do you think it is better to also change `Android Simulators` to `Android Emulators`?

File: MiniSim/MenuItems/SubMenuItem.swift (hunk 3 above)
Comment (itsramiel) [reply MiniSim/MenuItems/SubMenuItem.swift:174]: is it? the list is nested, so the bracket is indented. Does xcode remove the extra indentation for you when you format?

File: MiniSim/Service/DeviceService.swift (hunk 4 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:226]: I dont think it matters since we reset the value depending on whether `offline` is found or not.

I just didnt want to make it an optional and leaving it uninitialized makes xcode unhappy because not all paths set it

File: MiniSim/Service/DeviceService.swift (hunk 6 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:485]:
```suggestion
      splitted.removeFirst() // removes 'List of devices attached'
      let filtered = splitted.filter { !$0.contains("emulator") }
```

File: MiniSim/Service/DeviceService.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:44]: I dont mind doing that but I here is my opinion:

We are interested in the system uuid. Using `system_profiler` is implementation detail. If we ever changed that implementation we'll need to change the class name.

I would rather have a `System` class with a `currentUUID` and simply not care how it is implemented, but again do we need a completely new class for one method or do we wait till we need another system info and then create the class.

Another feature I wanted to add is to know whether a connected device is connected by usb or not. So we can use the System class and get `currentConnectedDevicesIds` then I guess it is worth having a class.

Anyways let me know what you think and I am indifferent really

**Comment by itsramiel** [comment]
> I've compiled the app and for some reason I can see my macbook there:
> 
> Here is the output of `xctrace list devices`
> 
> ```
> ❯ xctrace list devices
> == Devices ==
> MacBook Pro (Oskar) (123)
> iPhone (Oskar) (17.6.1) (123)
> 
> == Devices Offline ==
> Apple Vision Pro (2.0) (123)
> ```
> 
> Screenshot 2024-08-31 at 14 48 15
> 
> It would be nice to support the naming convention of `iPhone (Name)` as this is the default naming in my country and probably others too.

Would be great if we can setup a unit test for parsing results, so you know nothing breaks when fixing for other cases. I am not familiar with swift unit testing but that's an idea

**Comment by itsramiel** [comment]
> Also I found this: `xcrun devicectl list devices` which shows the output for listing only devices. Maybe faster (as it doesn't list simulators)

That is a pretty good command that I didnt know about. One issue I see is that it doesnt provide the ios version 🙃.

File: MiniSim/MenuItems/DeviceListSection.swift (hunk 2 above)
Comment (okwasniewski) [reply MiniSim/MenuItems/DeviceListSection.swift:21]: Oh sure, let's do this!

Comment (itsramiel) [reply MiniSim/MenuItems/DeviceListSection.swift]: done

**Comment by okwasniewski** [comment]
> Would be great if we can setup a unit test for parsing results, so you know nothing breaks when fixing for other cases. I am not familiar with swift unit testing but that's an idea

Yeah, this is great idea. I can write tests for current implementation in separate PR and then you can add it for real devices

File: MiniSim/Service/DeviceService.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (okwasniewski) [reply MiniSim/Service/DeviceService.swift:44]: Yeah true, let's abstract away the implementation details and name it System so even when we change commands underneath we don't need to rename. Good feature idea!

**Comment by okwasniewski** [comment]
I also learned about devicectl today, really handy and definitely bunch of cool stuff to implement in upcoming PRs! 

I also noticed that you can pass -j flag and specify json file to output the results. On the other hand not having version near the device is not such a big issue because you can't change the version anyways.

For simulators is much more useful because you can change their versions.

File: MiniSim/Service/DeviceService.swift (hunk 6 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift]: done

File: MiniSim/Service/DeviceService.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:44]: Well, if we'll use the other command to get the devices then we done even need this now 👍

**Comment by okwasniewski** [comment]
@itsramiel Just checked and it looks like Xcode 15 introduced `devicectl` in that case let's stay with your current approach. We can add features that use `devicectl` in upcoming PRs and check if user has Xcode 15 but let's add real device listing for everyone. 

Im going to write tests for parsers today.

**Comment by itsramiel** [comment]
> @itsramiel Just checked and it looks like Xcode 15 introduced `devicectl` in that case let's stay with your current approach. We can add features that use `devicectl` in upcoming PRs and check if user has Xcode 15 but let's add real device listing for everyone.
> 
> Im going to write tests for parsers today.

I just pushed the devicectl changes 😅

**Comment by okwasniewski** [comment]
@itsramiel Ugh, sorry.. okay i guess let's make this feature available only for Xcode >= 15. 

We will need to make sure that it doesn't break on older Xcode versions (if `xcrun devicectl` doesn't exist lets just don't display the iOS Devices section)

File: MiniSim/Service/DeviceService.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (okwasniewski) [reply MiniSim/Service/DeviceService.swift:44]: This is great! Let's use the `devicectl` command and make this feature Xcode 15 >= only. It's not going to do a lot without this command anyways (only display) so I think its reasonable to make this version lock

File: MiniSim/MenuItems/SubMenuItem.swift (hunk 3 above)
Comment (okwasniewski) [reply MiniSim/MenuItems/SubMenuItem.swift:174]: Oh okay I missed that you added additional indentation. nevermind

**Comment by okwasniewski** [comment]
I've extracted DeviceParser to another class and wrote some tests here: https://github.com/okwasniewski/MiniSim/pull/125

File: MiniSim/Service/DeviceService.swift [hunk 8]
```diff
@@ -263,14 +308,96 @@ extension DeviceService {
...
+      let connectionProperties: ConnectionProperties
+    }
+
+    struct Result: Codable {
+      let devices: [Device]
+    }
+
+    let info: Info
+    let result: Result
+  }
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:366]: I think this can be moved to DeviceParsers (the new class I've added) we can overload the file buffer there and do all decoding

Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift]: How would you do the overloading?

Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift]: @okwasniewski can you check my pushed changes if that matches what you mean?

File: MiniSim/Service/DeviceService.swift [hunk 9]
```diff
@@ -235,7 +279,45 @@ extension DeviceService {
 
   static func getIOSDevices() throws -> [Device] {
     Thread.assertBackgroundThread()
+    let simulators = try getIOSSimulators()
+    let devices = try getIOSPhysicalDevices()
+    return simulators + devices
+  }
+
+  static var currentMachineUUID: String? {
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:287]: This is not needed anymore, right?

File: MiniSim/Service/DeviceService.swift [hunk 10]
```diff
@@ -235,7 +279,45 @@ extension DeviceService {
 
   static func getIOSDevices() throws -> [Device] {
     Thread.assertBackgroundThread()
+    let simulators = try getIOSSimulators()
+    let devices = try getIOSPhysicalDevices()
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:283]: This can fail if user doesn't have Xcode 15. 

Even if the physical devices fail we want to display simulators

File: MiniSimTests/DeviceParserTests.swift [hunk 11]
```diff
@@ -105,18 +108,73 @@ class DeviceParserTests: XCTestCase {
...
+    XCTAssertEqual(devices[1].type, .virtual)
 
     XCTAssertEqual(devices[2].name, "Nexus_5X_API_28")
     XCTAssertEqual(devices[2].identifier, nil)
     XCTAssertFalse(devices[2].booted)
     XCTAssertEqual(devices[2].platform, .android)
+    XCTAssertEqual(devices[2].type, .virtual)
   }
 
+  func testAndroidPhysicalDeviceParser() {
```
Comment (okwasniewski) [inline MiniSimTests/DeviceParserTests.swift:126]: This is great! We are only missing the iOS physical devices test

File: MiniSim/Service/DeviceParser.swift [hunk 12]
```diff
@@ -50,6 +57,58 @@ class IOSSimulatorParser: DeviceParser {
   }
 }
 
+class IOSPhysicalDeviceParser: DeviceParser {
+  struct IOSPhysicalDevicesJson: Codable {
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceParser.swift:61]: Let's move this to the `Model` folder SwiftLint is complaining about this being nested.

File: MiniSim/Service/DeviceService.swift [hunk 13]
```diff
@@ -213,6 +214,49 @@ class DeviceService: DeviceServiceProtocol {
 
 // MARK: iOS Methods
 extension DeviceService {
+  private static func parseIOSPhysicalDevices(result: [String]) -> [Device] {
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:217]: This can also be removed

File: MiniSimTests/DeviceParserTests.swift [hunk 14]
```diff
@@ -105,18 +108,73 @@ class DeviceParserTests: XCTestCase {
...
+    XCTAssertEqual(devices.count, 1)
+
+    XCTAssertEqual(devices[0].name, "SM_A346E")
+    XCTAssertEqual(devices[0].identifier, "RFCWA0FXXXX")
+    XCTAssertTrue(devices[0].booted)
+    XCTAssertEqual(devices[0].platform, .android)
+    XCTAssertEqual(devices[0].type, .physical)
+  }
+
+
```
Comment (okwasniewski) [inline MiniSimTests/DeviceParserTests.swift:177]: Please run `swiftlint lint --fix ./` in the root to fix these linting issues

File: MiniSim/Service/DeviceService.swift [hunk 15]
```diff
@@ -359,7 +441,19 @@ extension DeviceService {
 
   static func getAndroidDevices() throws -> [Device] {
     Thread.assertBackgroundThread()
+    let emulators = try getAndroidEmulators()
+    let devices = try getAndroidPhysicalDevices()
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:445]: Same here, if physical devices fail we want to display emulators

**Review by okwasniewski — COMMENTED** [review]
We are getting there! Some additional cleanups are needed but overall it looks good

File: MiniSimTests/DeviceParserTests.swift (hunk 11 above)
Comment (itsramiel) [reply MiniSimTests/DeviceParserTests.swift:126]: Do we really need to test ios physical devices? It is a json result and we are not really parsing a raw string with regex but rather structured json. What would you test there?

File: MiniSim/Service/DeviceParser.swift (hunk 12 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceParser.swift:61]: I will but I believe that this struct is an implementation detail of the parsing of ios physical devices and need to stay private to it, perhaps add the private keyword.

File: MiniSim/Service/DeviceService.swift (hunk 13 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:217]: My bad, done

File: MiniSim/Service/DeviceService.swift (hunk 9 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:287]: yes, removed

File: MiniSim/Service/DeviceService.swift (hunk 10 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:240]: Would a guard statement with optional try and returning an empty array be good here?:
https://github.com/okwasniewski/MiniSim/blob/fa3503d984af87546eb980adca7dba38e0ea08db/MiniSim/Service/DeviceService.swift#L248-L251

File: MiniSimTests/DeviceParserTests.swift (hunk 14 above)
Comment (itsramiel) [reply MiniSimTests/DeviceParserTests.swift]: done

File: MiniSim/Service/DeviceService.swift (hunk 15 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:382]: Check my other comment to make sure it is what you want

File: MiniSim/Service/DeviceService.swift (hunk 10 above)
Comment (okwasniewski) [reply MiniSim/Service/DeviceService.swift:240]: Yeah I think this should be good

File: MiniSimTests/DeviceParserTests.swift (hunk 11 above)
Comment (okwasniewski) [reply MiniSimTests/DeviceParserTests.swift:126]: So we are making an assumption that the devicectl command will return this json every time but if the implementation changes or output of the command changes we will catch that in the tests.

I would love to cover MiniSim with more tests and this test will cover all the parsers

File: MiniSim/Service/DeviceParser.swift (hunk 12 above)
Comment (okwasniewski) [reply MiniSim/Service/DeviceParser.swift:61]: Yeah that's probably a better idea let's disable this warning

File: MiniSim/Service/DeviceService.swift (hunk 15 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:384]: done

File: MiniSim/Service/DeviceParser.swift (hunk 12 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceParser.swift:61]: done

File: MiniSim/Service/DeviceService.swift (hunk 10 above)
Comment (itsramiel) [reply MiniSim/Service/DeviceService.swift:240]: done

File: MiniSimTests/DeviceParserTests.swift (hunk 11 above)
Comment (itsramiel) [reply MiniSimTests/DeviceParserTests.swift:126]: Alright will try to find some time for it

Comment (itsramiel) [reply MiniSimTests/DeviceParserTests.swift:126]: Done

File: MiniSim/Service/DeviceService.swift [hunk 16]
```diff
@@ -39,6 +39,7 @@ class DeviceService: DeviceServiceProtocol {
   private enum ProcessPaths: String {
     case xcrun = "/usr/bin/xcrun"
     case xcodeSelect = "/usr/bin/xcode-select"
+    case systemProfiler = "/usr/sbin/system_profiler"
```
Comment (okwasniewski) [inline MiniSim/Service/DeviceService.swift:42]:
```suggestion
```

**Comment by okwasniewski** [comment]
Thanks for working on this!
